pubsublite/ps/settings.go

```diff
@@ -0,0 +1,187 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package ps
+
+import (
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/pubsublite/internal/wire"
+
+	pb "google.golang.org/genproto/googleapis/cloud/pubsublite/v1"
+)
+
+const (
+	// MaxPublishRequestCount is the maximum number of messages that can be
+	// batched in a single publish request.
+	MaxPublishRequestCount = wire.MaxPublishRequestCount
+
+	// MaxPublishMessageBytes is the maximum allowed serialized size of a single
+	// Pub/Sub message in bytes.
+	MaxPublishMessageBytes = wire.MaxPublishMessageBytes
+
+	// MaxPublishRequestBytes is the maximum allowed serialized size of a single
+	// publish request (containing a batch of messages) in bytes.
+	MaxPublishRequestBytes = wire.MaxPublishRequestBytes
+)
+
+// KeyExtractorFunc is a function that extracts an ordering key from a Message.
+type KeyExtractorFunc func(*pubsub.Message) []byte
+
+// PublishMessageTransformerFunc transforms a pubsub.Message to a Pub/Sub Lite
+// PubSubMessage. If this returns an error, the pubsub.PublishResult will be
+// errored and the PublisherClient will consider this a fatal error and
+// terminate.
+type PublishMessageTransformerFunc func(*pubsub.Message) (*pb.PubSubMessage, error)
+
+// PublishSettings control the batching of published messages. These settings
+// apply per partition.
+//
+// Use DefaultPublishSettings for defaults, as an empty PublishSettings will
+// fail validation.
+type PublishSettings struct {
+	// Publish a non-empty batch after this delay has passed. Must be > 0.
+	DelayThreshold time.Duration
+
+	// Publish a batch when it has this many messages. Must be > 0. The maximum is
+	// MaxPublishRequestCount.
+	CountThreshold int
+
+	// Publish a batch when its size in bytes reaches this value. Must be > 0. The
+	// maximum is MaxPublishRequestBytes.
+	ByteThreshold int
+
+	// The maximum time that the client will attempt to establish a publish stream
+	// connection to the server. Must be > 0.
+	//
+	// The timeout is exceeded, the publisher will terminate with the last error
+	// that occurred while trying to reconnect. Note that if the timeout duration
+	// is long, ErrOverflow may occur first.
+	Timeout time.Duration
+
+	// The maximum number of bytes that the publisher will keep in memory before
+	// returning ErrOverflow. Must be > 0.
+	//
+	// Note that Pub/Sub Lite topics are provisioned a publishing throughput
+	// capacity, per partition, shared by all publisher clients. Setting a large
+	// buffer size can mitigate transient publish spikes. However, consistently
+	// attempting to publish messages at a much higher rate than the publishing
+	// throughput capacity can cause the buffers to overflow. For more
+	// information, see https://cloud.google.com/pubsub/lite/docs/topics.
```

Comment:
Is this something that Java/Python do, too?

Reply:
Not Java. I haven't looked at Python closely, but I don't remember seeing it implemented there either.

It's for feature parity with the CPS Go client's [pubsub.PublishSettings.BufferedByteLimit](https://godoc.org/cloud.google.com/go/pubsub#PublishSettings).

Reply:
We added Publisher flow control in Pub/Sub for [Java](https://github.com/googleapis/java-pubsub/pull/119/files) and I think Python too. I'm currently adding this for Go, which will add `MaxOutstandingMessages` and `MaxOutstandingBytes`. We should have a discussion on how we're going to keep feature parity between the two libraries.

Reply:
Definitely; I've made a note to start a discussion once everyone is back from holidays.

P.S. Thanks for all the reviews!